File: examples/js/js-student_grades-02_extract_subclass/src/StudentGradeCalculator.js
export class StudentGradeCalculator {
    allYearsTeachers = {
        2020: [
            ["Josefina", true],
            ["Edonisio", true],
            ["Edufasio", false],
        ],
        2019: [
            ["Eduarda", false],
            ["Abelardo", false],
            ["Francisca", false],
        ]
    };
    
    constructor(yearToCalculate) {
        this.yearToCalculate = yearToCalculate;
    }
    
    calculate(examGrades, hasReachedMinimumClasses) {
        if (examGrades.length !== 0) {
            let hasToIncreaseOneExtraPoint = false;
            
            for (let [year, teachers] of Object.entries(this.allYearsTeachers)) {
                if (!(this.yearToCalculate != year)) {
                    for (let teacher of teachers) {
                        if (teacher[1] != true) {
                            continue;
                        }
                        
                        hasToIncreaseOneExtraPoint = true;
                    }
                } else {
                    continue;
                }
            }
            
            let gradesSum       = 0;
            let gradesWeightSum = 0;
    
            for (let examGrade of examGrades) {
                gradesSum += (examGrade[0] * examGrade[1] / 100);
                gradesWeightSum += examGrade[0];
            }
            
            if (gradesWeightSum === 100) {
                if (hasReachedMinimumClasses) {
                    if (hasToIncreaseOneExtraPoint) {
                        return Math.min(10, gradesSum + 1);
                    } else {
                        return gradesSum;
                    }
                } else {
                    return 0;
                }
            } else if (gradesWeightSum > 100) {
                return -1;
            } else {
                return -2;
            }
        } else {
            return 0;
        }
    }
}

File: examples/java/java-student_grades-04_replace_control_flag/src/main/java/tv/codely/student_grades/StudentGradeCalculator.java
package tv.codely.student_grades;

import java.util.AbstractMap;
import java.util.List;
import java.util.Map;

public class StudentGradeCalculator {
    final private Map<Integer, List<Pair<String, Boolean>>> allYearsTeachers = Map.ofEntries(
        new AbstractMap.SimpleImmutableEntry<>(
            2020,
            List.of(
                new Pair<>("Josefina", true),
                new Pair<>("Edonisio", true),
                new Pair<>("Edufasio", false)
            )
        ),
        new AbstractMap.SimpleImmutableEntry<>(
            2019,
            List.of(
                new Pair<>("Eduarda", false),
                new Pair<>("Abelardo", false),
                new Pair<>("Francisca", false)
            )
        )
    );
    private final int                                       yearToCalculate;

    public StudentGradeCalculator(final int yearToCalculate) {
        this.yearToCalculate = yearToCalculate;
    }

    public float calculateGrades(final List<Pair<Integer, Float>> examsGrades, final boolean hasReachedMinimumClasses) {
        if (examsGrades.isEmpty()) {
            return 0f;
        }

        if (!hasReachedMinimumClasses) {
            return 0f;
        }

        ensureGradesWeightSumIs100percent(examsGrades);

        float gradesSum = gradesSum(examsGrades);

        if (hasToIncreaseOneExtraPoint()) {
            return Float.min(10f, gradesSum + 1);
        }

        return gradesSum;
    }

    private void ensureGradesWeightSumIs100percent(List<Pair<Integer, Float>> examsGrades) {
        int gradesWeightSum = gradesWeightSum(examsGrades);

        if (gradesWeightSum > 100) {
            throw new GradesWeightOverMaxException();
        } else if (gradesWeightSum < 100) {
            throw new GradesWeightBelowMinException();
        }
    }

    private float gradesSum(List<Pair<Integer, Float>> examsGrades) {
        float gradesSum = 0f;
        for (Pair<Integer, Float> examGrade : examsGrades) {
            gradesSum += (examGrade.first() * examGrade.second() / 100);
        }
        return gradesSum;
    }

    private int gradesWeightSum(List<Pair<Integer, Float>> examsGrades) {
        return examsGrades.stream().map(Pair::first).reduce(Integer::sum).get();
    }

    private boolean hasToIncreaseOneExtraPoint() {
        boolean hasToIncreaseOneExtraPoint = false;

        for (Map.Entry<Integer, List<Pair<String, Boolean>>> yearlyTeachers : allYearsTeachers.entrySet()) {
            if (!(yearToCalculate != yearlyTeachers.getKey())) {
                List<Pair<String, Boolean>> teachers = yearlyTeachers.getValue();

                for (Pair<String, Boolean> teacher : teachers) {
                    if (teacher.second() != true) {
                        continue;
                    }
                    hasToIncreaseOneExtraPoint = true;
                }
            } else {
                continue;
            }
        }
        return hasToIncreaseOneExtraPoint;
    }
}
